### ISSUE TITLE
Dealing with stdlib JLLs in dependency graphs

### ISSUE THREAD

**Issue by staticfloat** [JuliaPackaging/Yggdrasil#1682]
There are some very annoying consquences around having stdlib JLLs.  Quick example:

* Julia 1.5 ships with GMP 6.1.2, Julia 1.6 ships with GMP 6.2.0
* `Singular_jll` links against GMP, and libtool embeds a `-compatibility_version` flag into `gcc` that is minor-version-specific, meaning that `Singular_jll` _must_ have 6.2.0 or newer.
* While we can use JLL `Project.toml` compat entries to select the appropriate JLL for a Julia version, it may be difficult to get this to work for dependencies such as `Singular_jll`.

I can see a few ways we could improve the situation:

* Build `Singular_jll` against an older version of GMP (e.g. 6.1.2) and then use that on 6.2.0 (this should work)
* Coordinate `Singular_jll` and `GMP_jll` versions by creating a `Singular_jll` that works with `GMP_jll 6.*`, then another `Singular_jll` that works only with `GMP_jll 6.2.*`.  The hardest part there will be creating a `Singular_jll` that is built against `GMP_jll 6.1.2` but declares a compat entry against `GMP_jll 6.*`.
* Note that on Julia 1.6+, with the `julia_version` tag on `Platform` objects, we can actually have a single release of `Singular_jll` that has artifacts for Julia 1.5 and artifacts for Julia 1.6.  I'm not crazy about this idea though, as I'm not 100% sure whether to put intelligence into `Project.toml` or `Artifacts.toml` as far as choosing the right Singular to use.

**Comment by vchuravy** [comment]
It feels odd to put it in Project.toml since it is not an actual
compatibility restriction of the software, but rather it is a compatibility
restriction of the produced binary.

But on the other hand we use the version of the jll to encode things like
that. We have similar issues with building against LLVM. I can build one
and the same version of Enzyme against 5 different LLVM versions.

(This is getting dangerously close to the ABI issues that we were talking
about with MPI)

**Comment by giordano** [comment]
> * Build Singular_jll against an older version of GMP (e.g. 6.1.2) and then use that on 6.2.0 (this should work)

This is an ad-hoc solution for GMP downstream packages.  I believe we should find a more general solution, since this keeps biting us.

> * Note that on Julia 1.6+, with the julia_version tag on Platform objects, we can actually have a single release of Singular_jll that has artifacts for Julia 1.5 and artifacts for Julia 1.6. I'm not crazy about this idea though, as I'm not 100% sure whether to put intelligence into Project.toml or Artifacts.toml as far as choosing the right Singular to use.

I start feeling like we should have all binary dependencies of julia (including LLVM!) as part of the platform and build against multiple versions of them when they are actual dependencies of the current package.  This goes into the same direction as putting intelligence into Artifacts.toml, but this is a fairly crazy idea and I currently have no clue how we could deal with this in older Julia versions.

**Comment by staticfloat** [comment]
> > * Build Singular_jll against an older version of GMP (e.g. 6.1.2) and then use that on 6.2.0 (this should work)
> This is an ad-hoc solution for GMP downstream packages. I believe we should find a more general solution, since this keeps biting us.

True, but the generalized solution is not too crazy; it's a decoupling of build-time dependency version specification and runtime dependency version specification.  Right now, if we declare that `Singular_jll` has a dependency on GMP and it's okay with GMP 6.1.2-6.2, then it will build with 6.2.0 because our runtime and build-time resolution algorithms are the same; they both pick the latest possible one.  A generalized solution for the problem of `-compatibility_version` linker flags is to force building against an older version while allowing installation of newer versions at runtime.

> It feels odd to put it in Project.toml since it is not an actual compatibility restriction of the software, but rather it is a compatibility restriction of the produced binary.

This is an interesting distinction, but since the only resolver we have is the Pkg one, I think the distinction is all but meaningless for us.

This has helped me to narrow down my thinking a bit:

Any compatibility constraints we want to express that need to effect the global resolution _must_ be expressed at the `Project.toml` level.  The `Artifacts.toml` level can be used to further refine the match, but should be reserved for things that are out of the control of the resolver.

The reason this issue is so thorny is because thanks to the binaries shipped with Julia, we have a tying together of something that is outside of the resolver's control (julia version) and something that ordinarily should be in the resolver's control (package compatibility).  I think a good chunk of this will be solved by having JLL stdlibs, (because then the resolver will know that GMP 6.1.2 is already installed) and then making our version dependencies tighter (e.g. having builds of `Singular_jll` that build against GMP 6.1.2 that are then compatible with a range from 6.1.2+).  This way we can more effectively utilize package compatibility to express what it's best at, which is the actual compatibility of the binaries within the JLLs themselves.

This does _not_ solve the issue of things like dependencies that are tightly-integrated with LLVM.  In this case, we have a true ABI compat issue, and while we can typically dodge this by saying "everybody gets the same, latest version of dependency X" that obviously doesn't work with stdlibs that don't get updates.  In these cases, (GMP, MPFR, LLVM, PCRE2, LibCURL, etc...) I think we're just going to have to bite the bullet and either maintain multiple JLL versions that declare different compat bounds on these stdlib JLLs, or we're going to have to build multiple artifacts and shard them based on `julia_version` so that the appropriate one gets selected at `Pkg.add()` time.  I think that should work for things like `Enzyme`.

**Comment by giordano** [comment]
> True, but the generalized solution is not too crazy; it's a decoupling of build-time dependency version specification and runtime dependency version specification.  Right now, if we declare that `Singular_jll` has a dependency on GMP and it's okay with GMP 6.1.2-6.2, then it will build with 6.2.0 because our runtime and build-time resolution algorithms are the same; they both pick the latest possible one.  A generalized solution for the problem of `-compatibility_version` linker flags is to force building against an older version while allowing installation of newer versions at runtime.

#1679 is a case where the problem isn't a different compatibility version but new features enabled in the latest version of MbedTLS.  I'm not convinced that building with an older binary will always solve our issues.

I just wish we didn't have clashes at all with Julia shipped libraries (who cares if GMP v6.2 is loaded in Julia v1.5 by a third-party library?  Its their duty to use the right API, Julia isn't going to call into that one directly), that would solve tons of headaches...

**Comment by fingolfin** [comment]
> Build Singular_jll against an older version of GMP (e.g. 6.1.2) and then use that on 6.2.0 (this should work)

Yes, that would work, and it would be what I'd prefer. But one major issue with that is that it cannot be done on Yggdrasil right now, because that runs Julia 1.6 and as such it is impossible to build a recipe against GMP_jll 6.1.2 (at least I believe that was the case last Friday when I "accidentally" tried this as part of the libjulia_jll work).


As to loading GMP 6.2 in Julia 1.5: None of the packages that I am involved with use any of the super minor new APIs in GMP 6.2. However, what *is* done (even by Julia itself!) is that during initialization, some GMP functions are called to set custom memory allocation handlers in GMP. If one now loads a newer GMP 6.2, then down the road we may end up calling GMP 6.2 functions on objects allocated with GMP 6.1, and that will blow. Actually, this whole scenario was a major driving force for adding `init_block` in https://github.com/JuliaPackaging/BinaryBuilder.jl/pull/791. So if one adapted GMP_jll to use `init_block` suitably, then *perhaps* it would be safe to use 6.2 with older Julia versions?

Well, except for the reverse direction: if an object allocated by GMP 6.2 is then passed to GMP 6.1 code, it might again run into an issue: because one thing GMP 6.2 added is lazy allocation of `mpz_t` objects; there a null pointer can be used in a way where in GMP 6.1 this would result in a crash.

All in all, I don't feel very confident about loading multiple GMP versions into a single process...

**Comment by staticfloat** [comment]
With https://github.com/JuliaPackaging/BinaryBuilderBase.jl/commit/cab650c82efd15cf5676bc05b553c5fbcb7f1054 we have a couple of new capabilities.

First, if your `Platform` object has a `julia_version` embedded within it, it will automatically resolve build versions with respect to that julia version.  So you can set a `Platform("x86_64", "linux"; julia_version=v"1.5")` and you'll always get GMP 6.1.2.

If you don't specify any `julia_version`, it will completely ignore julia version compatibility; you'll just get the default resolver choice of versions, which is usually the latest possible given your compat bounds.

If you want to build with a specific version, but don't want to artificially limit your choices of compatibility at runtime, you can use the new `build_version` argument to `Dependency()`.  In short, you can now say:

```
dependencies = Dependency[
    Dependency("GMP_jll", v"6.1.2"),
]
```

This means, "use `v"6.1.2"` at build time, and express no runtime compat bound".  If you want a compat bound with a build-time version set as well, you still need to use the `PkgSpec` stuff.

**Comment by fingolfin** [comment]
So having `julia_version` in the platform control how packages are select was super helpful for `libjulia_jll`. And there, we don't want `julia_version` to affect the platform strings or which artifacts to choose, it is only there to control how dependencies are resolved.

But there is still need for a way to have a `julia_version` variant which can be used to produce multiple artifacts, for different Julia versions. This is e.g. need for `libcxxwrap_julia` and the 10 or so JLLs building atop it: they want one binary for use in Julia 1.3, linked against libjulia_jll 1.3, and with `julia_compat` restricted to `~1.3`;  and (at least) one more set of artifacts for Julia >= 1.4 (built against libjulia_jll 1.4 and  `julia_compat` restricted to `^1.4`;  though to be on the safe side, it'd perhaps be even better to have separate artifacts for each of Julia 1.3, 1.4, 1.5, 1.6, ... (in case there are other relevant ABI changes that we are not yet aware).

Of course one can also try to hack around this, say by having libcxxwrap version   0.10.x built against julia 1.3 (and with the corresponding julia_compat),  but version 0.11.x built against libjulia 1.4, etc. but that quickly turns into a major hassle if one needs/wants to support more than one Julia version at a time...
